Feature/92 logging

IIIjong opened the pull request:
## 📝 작업 내용 (What I did)

**관리자 관점의 로그 관측 및 감사(Audit) 기능 기반을 구축하였다.**
Spring Boot 기반 서비스에서 발생하는 요청·에러·행위 로그를 **Elasticsearch에 적재**하고, 관리자 API를 통해 **조건 검색 및 페이징 조회가 가능하도록 구성하였다.**

* **API Log**: 컨트롤러 호출 기준으로 요청 메서드, URI, 응답 시간, 사용자, 클라이언트 IP를 기록
* **Error Log**: 예외 발생 시 예외 타입 및 메시지를 저장하여 장애 분석이 가능하도록 구성
* **Action Log**: 주요 도메인 행위(CRUD)에 대해 Before/After 스냅샷 기반의 감사 로그를 기록 (AOP + 커스텀 어노테이션)

또한 관리자 조회 응답 DTO 단계에서 **IP 마스킹 처리**를 적용하여 민감 정보 노출을 최소화하였다.

---

## ✨ 변경 사항 (Changes)

* [x] 주요 기능 추가 / 변경
* [x] 코드 리팩토링 (검색 조건 및 응답 DTO 구조화)
* [ ] 버그 수정
* [ ] 기타

---

## 🔀 작업한 브랜치 (Working Branch)

* `feature/92-logging`

---

## #️⃣ 관련 이슈 (Issue)

* Closes #92 

---

## ✨ 상세 변경 내용

### 1) Elasticsearch 기반 로그 도메인 및 Repository 구성

* `ApiLog`, `ErrorLog`, `ActionLog`를 Elasticsearch `@Document`로 분리하여 구성하였다.
* 로그 유형별 `ElasticsearchRepository`를 각각 정의하였다.

  * `ApiLogRepository`
  * `ErrorLogRepository`
  * `ActionLogRepository`
* `ElasticsearchConfig`에서 로그 Repository 패키지를 명시적으로 스캔하도록 설정하였다.

**설계 의도**
운영 로그, 장애 로그, 감사 로그를 명확히 분리하여 인덱스 전략 및 조회 정책을 독립적으로 확장할 수 있도록 설계하였다.

---

### 2) 관리자 로그 조회 API 구현 (검색 조건 + 페이징)

관리자 권한(ADMIN)을 대상으로 다음 조회 API를 구성하였다.

* `/api/v1/logs/apis` : API 로그 조회
* `/api/v1/logs/errors` : 에러 로그 조회
* `/api/v1/logs/actions` : 행위 로그 목록 조회
* `/api/v1/logs/actions/{id}` : 행위 로그 상세 조회 (Before/After 스냅샷 포함)

검색 조건 DTO를 로그 유형별로 분리하였다.

* `AdminLogSearchCondition`
* `AdminErrorLogSearchCondition`
* `AdminActionLogSearchCondition`

API 로그 조회의 경우 `AdminApiLogSearchRepositoryImpl`에서
`CriteriaQuery` 기반 검색 로직과 페이징 처리를 구현하였다.

**설계 의도**
운영 환경에서 기본적으로 요구되는 **기간 기반 검색 + 조건 필터링 + 페이징** 요구사항을 충족하기 위함이다.

---

### 3) AOP 기반 자동 로깅 처리

#### (1) API Logging Aspect

* `Controller` 계층을 대상으로 실행 시간을 측정하였다.
* HTTP method, URI, 사용자(principal), 클라이언트 IP, 처리 시간을 기록하였다.
* 정상 흐름 기준(status=200) 로그를 Elasticsearch에 저장하도록 구성하였다.

#### (2) Action Log Aspect (감사 로그)

* `@ActionLoggable` 어노테이션을 기반으로 특정 행위를 자동 수집하도록 구성하였다.
* SpEL을 이용해 entityId를 추출하였다.
* `CREATE / UPDATE / DELETE` 행위에 대해 Before/After 스냅샷을 JSON 형태로 저장하였다.
* 예외 발생 시 `success=false`, `failReason`에 예외 메시지를 기록하였다.
* `EntityManager.flush()` 이후 After 스냅샷을 확보하였다.

**리뷰 요청 포인트**

* SpEL 기반 entityId 추출 방식이 팀 표준에 적합한지
* ActionLog 스냅샷 범위(민감 필드 포함 여부)에 대한 개선 방향

---

### 4) 응답 DTO 기준 민감 정보(IP) 마스킹

* 관리자 조회 응답 DTO에서 `clientIp` 값을 마스킹 처리하였다.
* 실제 저장 데이터는 유지하고, **응답 레벨에서만 가공**하도록 구성하였다.

**의도**
관리자 화면 및 API 응답에서 불필요한 개인정보 노출을 방지하기 위함이다.

IIIjong committed:
refactor: log 패키지 제거

IIIjong committed:
setting: aop, elasticsearch 의존성 추가

IIIjong committed:
feat: Elasticsearch 기반 API 로그 수집 1차 구축

- Docker Compose 기반 Elasticsearch 실행
- Spring Boot - Elasticsearch 연동
- API 로그 도메인 및 Repository 구성
- 로그 저장 검증용 테스트 API 추가
- Controller/AOP 로깅 구조 설계

IIIjong committed:
refactor: Entity @Audited, @EntityListeners 어노테이션 제거

IIIjong committed:
[chore]: Envers(@Audited) 감사 어노테이션 제거

- Hibernate Envers 기반 _AUD 테이블 생성 방지
- 로그 시스템을 커스텀 로깅 구조로 전환하기 위한 정리 작업

IIIjong committed:
[feat]: 사용자 생성/수정/삭제에 액션 로그 연동

- @ActionLoggable 적용
- empNo 기반 엔티티 식별 방식 정리

IIIjong committed:
[feat]: Elasticsearch 기반 관리자 로그 시스템 구현

- API / Error / Action 로그 분리 저장
- AOP 기반 자동 로그 수집
- 변경 전·후 스냅샷 및 Diff 조회 기능 추가
- 관리자 검색 / 필터링 / 페이지네이션 API 제공

IIIjong committed:
merge: develop 병합

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/service/AdminActionLogService.java:
```diff
@@ -0,0 +1,35 @@
+            .orElseThrow(() -> new IllegalArgumentException("행위 로그를 찾을 수 없습니다."));
+
+        return ActionLogDetailResponse.from(log);
+    }
```
500에러 표기는 의도된건가요?

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/snapshot/UserSnapshot.java:
```diff
@@ -0,0 +1,45 @@
+                    : null
+            )
+            .build();
+    }
```
매개변수 final 명시 부탁드립니다~

IIIjong replied:
완료

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/user/UserService.java:
```diff
@@ -8,9 +8,9 @@
-    void deleteUser(Long userId);
+    void deleteUser(String enpNo);
     UserResponse findUserByEnpNo(String enpNo);
     void updateUser(String enpNo,UpdateUserRequest request);;
```
매개변수 final 명시 부탁드립니다~

IIIjong replied:
완료

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/user/User.java:
```diff
@@ -24,14 +22,11 @@
 @Entity
 @Getter
 @Builder
 @ToString
-@Audited
```
@ToString 도 지우기

IIIjong replied:
완료

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/aspect/ApiLoggingAspect.java:
```diff
@@ -0,0 +1,58 @@
+            .timestamp(LocalDateTime.now())
+            .build();
+
+        apiLogRepository.save(apiLog);
```
Http상태 코드가 201이나 204인 경우에는 어떻게 되나요?

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/controller/AdminActionLogController.java:
```diff
@@ -0,0 +1,60 @@
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/logs/actions")
+@RequiredArgsConstructor
+@PreAuthorize("hasRole('ADMIN')")
+public class AdminActionLogController {
```
PreAuthorize를 클래스 상단말고 각 메서드 마다 선언해주는 건 어떤가요?

IIIjong replied:
해당 컨트롤러가 어드민 전용이라는 것을 명시했긴때문에 저는 어노테이션 중복으로 메소드마다 각각 적을 필요없다고 생각해서 현상태 유지하겠습니다

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/controller/AdminApiLogController.java:
```diff
@@ -0,0 +1,49 @@
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/logs")
+@RequiredArgsConstructor
+@PreAuthorize("hasRole('ADMIN')")
+public class AdminApiLogController {
```
여기도 역할 설정 메서드마다는 어떤가요?

IIIjong replied:
위 코멘트와 동일

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/controller/AdminErrorLogController.java:
```diff
@@ -0,0 +1,49 @@
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/logs/errors")
+@RequiredArgsConstructor
+@PreAuthorize("hasRole('ADMIN')")
+public class AdminErrorLogController {
+
```
여기도 역할 마찬가지

IIIjong replied:
위 코멘트와 동일

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/dto/response/ApiLogResponse.java:
```diff
@@ -0,0 +1,39 @@
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+@Builder
+@AllArgsConstructor
+public class ApiLogResponse {
+
+    private final String id;
+    private final String timestamp;
+    private final String method;
+    private final String uri;
+    private final int status;
+    private final long duration;
+    private final String principal;
+    private final String clientIp;
+
+    public static ApiLogResponse from(ApiLog log) {
+        return ApiLogResponse.builder()
+            .id(log.getId())
+            .timestamp(log.getTimestamp().toString())
+            .method(log.getMethod())
+            .uri(log.getUri())
+            .status(log.getStatus())
+            .duration(log.getDuration())
+            .principal(log.getPrincipal())
+            .clientIp(maskIp(log.getClientIp()))
+            .build();
+    }
```
여기에서 timestamp는 기록된 시간대를 알려주는 거죠?

IIIjong replied:
맞습니다

wjdvlf5456 commented on smerp/src/main/java/com/domino/smerp/logging/aspect/ActionLogAspect.java:
```diff
@@ -0,0 +1,161 @@
+
+            em.flush();
+
+            if ("CREATE".equals(action)) {
+                context.setVariable("result", result);
+                entityId = parseSpel(actionLoggable.entityId(), context);
+            }
```
CREATE, DELETE 같은 값들 상수나 ENUM으로 관리하는 건 어떤가요?

IIIjong committed:
refactor: 팀 컨벤션에 따른 final 명시

IIIjong committed:
refactor: 팀 컨벤션에 따른 final 명시

IIIjong committed:
feat: API 로그 수집을 AOP에서 Filter 기반으로 전환

- Controller AOP 기반 API 로그 수집 제거
- OncePerRequestFilter 기반 ApiLoggingFilter 도입
- 실제 HTTP 응답 상태 코드(200/201/204/4xx/5xx) 정확히 기록
- ApiLog / ErrorLog 책임 분리
- 관리자 API 로그 조회 시 기본적으로 정상(status < 400) 요청만 노출

IIIjong committed:
refactor: ActionLog 엔티티 스냅샷 로딩을 Provider 패턴으로 리팩토링

- ActionLogAspect에서 엔티티별 분기 로직 제거
- ActionLogEntityProvider 인터페이스 도입
- Entity 타입별 스냅샷 조회 책임을 Provider 구현체로 분리
- Provider 자동 수집을 위한 ActionLogEntityProviderRegistry 추가
- USER 엔티티에 대한 UserActionLogProvider 구현
- 신규 엔티티 추가 시 Aspect 수정 없이 Provider 확장 가능하도록 구조 개선

wjdvlf5456 reviewed (dismissed):
확인했습니다.

IIIjong committed:
refactor: 액션 로그 예외 처리 코드 추가

wjdvlf5456 reviewed (approved):
예외처리 확인했습니다
